### PR TITLE
CupertinoPicker semantics

### LINKED ISSUE #20561: Make CupertinoPicker accessible
Edit:
Currently cupertino picker directly exposes the scrollable list of children. Instead it needs to describe it's semantics as an adjustable, with values set appropriately and all child semantics hidden.

_____
Because the CupertinoPicker can take an arbitrary list of widgets, determining the current, previous, and next accessibility labels for the adjustable semantics is non-trivial. Consider the following snippet of code to make this work currently:

```dart
class RenderCupertinoSemantics extends RenderProxyBox {
 ...
@override
  void assembleSemanticsNode(SemanticsNode node, SemanticsConfiguration config, Iterable<SemanticsNode> children) {
    if (children.isEmpty)
      return super.assembleSemanticsNode(node, config, children);
    final SemanticsNode scrollable = children.first;
    final List<SemanticsNode> labels = <SemanticsNode>[];
    scrollable.visitChildren((SemanticsNode child) {
      labels.add(child);
      return true;
    });
    int middle = 0;
    for (int i = 0; i < labels.length; i++) {
      final SemanticsNode label = labels[i];
      final double offset = label.transform.getTranslation()[1];
      if (offset > _kMagicMinOffset && offset < _kMagicMaxOffset) {
        middle = i;
      }
    }
    config.value = labels[middle].label;
    if (middle + 1 < labels.length)
      config.increasedValue = labels[middle + 1].label;
    if (middle - 1 >= 0)
      config.decreasedValue = labels[middle - 1].label;
    super.assembleSemanticsNode(node, config, const <SemanticsNode>[]);
  }
}

```

The only way to find the values is to traverse the set of built SemanticsNodes and use the scroll offset to guestimate which one came from the center. 

A data type like `CupertinoPickerItem` or `CupertinoPickerItem<T>` could also encapsulate the offsetting rules, while providing a getter which returned a string representation we could use for semantics. Then the above calculation would just be:

```dart
  @override
  void describeSemanticsConfiguration(SemanticsConfiguration config) {
    super.describeSemanticsConfiguration(config);
    config.isSemanticBoundary = true;
    config.value = child[_selectedItem].value.toString();
    if (_childCount != null && _selectedItem < _childCount - 1) {
      config.onIncrease = _handleIncrease;
      config.value = child[_selectedItem + 1].value.toString();
    }
    if (_selectedItem > 0) {
      config.onDecrease = _handleDecrease;
      config.value = child[_selectedItem-+ 1].value.toString();
    }
    config.textDirection = textDirection;
  }
```

cc @matthew-carroll , @xster

### PR DESCRIPTION
**Pull request by jonahwilliams**
Make Cupertino picker accessible. Added simple test and verified with a11y inspector on simulator.

Turns the scroll semantics of the ListView into a single adjustable semantics  node. This is done by removing all of the child semantics of the scroll wheel and using the scroll indexes to loop up the current, previous, and next semantic label. This label is then turned into the value of a new adjustable semantic node, with adjustment callbacks wired to move the scroll controller.

Fixes https://github.com/flutter/flutter/issues/20561

Combined results from https://github.com/flutter/flutter/pull/20653 and  https://github.com/flutter/flutter/pull/21303

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/picker.dart [hunk 1]
```diff
@@ -274,3 +284,110 @@ class _CupertinoPickerState extends State<CupertinoPicker> {
...
+
+  @override
+  void assembleSemanticsNode(SemanticsNode node, SemanticsConfiguration config, Iterable<SemanticsNode> children) {
+    if (children.isEmpty)
+      return super.assembleSemanticsNode(node, config, children);
+    final SemanticsNode scrollable = children.first;
+    final Map<int, SemanticsNode> indexedChildren = <int, SemanticsNode>{};
+    scrollable.visitChildren((SemanticsNode child) {
+      assert(child.indexInParent != null);
+      indexedChildren[child.indexInParent] = child;
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/picker.dart]: Should we document somewhere that semantics for this widget only work well if children are wrapped in IndexedSemantics?

Comment (goderbauer) [reply packages/flutter/lib/src/cupertino/picker.dart]: Actually, never mind. Looks like the CupertinoPicker is always in charge of instantiating the delegate, so this is a purely internal issue.

**Review by goderbauer — APPROVED** [review]
LGTM with one request for an extra test.

File: packages/flutter/lib/src/cupertino/picker.dart [hunk 2]
```diff
@@ -161,6 +161,13 @@ class CupertinoPicker extends StatefulWidget {
 
 class _CupertinoPickerState extends State<CupertinoPicker> {
   int _lastHapticIndex;
+  FixedExtentScrollController _controller;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = widget.scrollController ?? FixedExtentScrollController();
+  }
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/picker.dart]: You'd need didUpdateWidget here too no?

File: packages/flutter/test/cupertino/date_picker_test.dart [hunk 3]
```diff
@@ -581,4 +583,54 @@ void main() {
...
\ No newline at end of file
+
+  testWidgets('CupertinoPicker exports semantics', (WidgetTester tester) async {
+    final SemanticsHandle handle = tester.ensureSemantics();
+    await tester.pumpWidget(
+      Directionality(
+        textDirection: TextDirection.ltr,
+        child: CupertinoPicker(
+          itemExtent: 100.0,
+          onSelectedItemChanged: (int index) {},
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/test/cupertino/date_picker_test.dart]: Done

File: packages/flutter/lib/src/cupertino/picker.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/cupertino/picker.dart]: Added didUpdateWidget and a test case to confirm that this works properly

**Pull request merged**